xtm/xtm_api.cc

```diff
@@ -0,0 +1,354 @@
+		malloc(sizeof(struct xtm_queue) +
+		       size * sizeof(union xtm_msg));
+	if (queue == NULL)
+		return NULL;
+
+	queue->flags = flags;
+	queue->is_notification_expected = false;
+
+	if (create_fds(&queue->consumer_read_fd,
+		       &queue->consumer_write_fd) < 0) {
+		save_errno = errno;
+		goto free_queue;
+	}
+
+	if (queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) {
+		if (create_fds(&queue->producer_read_fd,
+			       &queue->producer_write_fd) < 0) {
+			save_errno = errno;
+			goto close_consumer_fds;
+		}
+	} else {
+		queue->producer_read_fd = queue->producer_write_fd = -1;
+	}
+
+	if (fcntl(queue->consumer_read_fd, F_SETFL, O_NONBLOCK) < 0 ||
+	    fcntl(queue->consumer_write_fd, F_SETFL, O_NONBLOCK) < 0 ||
+	    (queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS &&
+	    (fcntl(queue->producer_read_fd, F_SETFL, O_NONBLOCK) < 0 ||
+	     fcntl(queue->producer_write_fd, F_SETFL, O_NONBLOCK) < 0))) {
+		save_errno = errno;
+		goto close_producer_fds;
+	}
+	if (queue->queue.create(size) < 0) {
+		save_errno = EINVAL;
+		goto close_producer_fds;
+	}
+	return queue;
+
+close_producer_fds:
+	if (queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) {
+		close(queue->producer_read_fd);
+		if (queue->producer_read_fd != queue->producer_write_fd)
+			close(queue->producer_write_fd);
+	}
+close_consumer_fds:
+	close(queue->consumer_read_fd);
+	if (queue->consumer_read_fd != queue->consumer_write_fd)
+		close(queue->consumer_write_fd);
+free_queue:
+	free(queue);
+	errno = save_errno;
+	return NULL;
+};
+
+int
+xtm_queue_delete(struct xtm_queue *queue)
+{
+	int rc = 0;
+	if ((queue->flags & XTM_QUEUE_NEED_TO_CLOSE_READFD) &&
+	    (queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) &&
+	    close(queue->producer_read_fd) < 0)
+		rc = -1;
+	if ((queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) &&
+	    queue->producer_write_fd != queue->producer_read_fd &&
+	    close(queue->producer_write_fd) < 0)
+		rc = -1;
+	if ((queue->flags & XTM_QUEUE_NEED_TO_CLOSE_READFD) &&
+	    close(queue->consumer_read_fd) < 0)
+		rc = -1;
+	if (queue->consumer_read_fd != queue->consumer_write_fd &&
+	    close(queue->consumer_write_fd) < 0)
+		rc = -1;
+	free(queue);
+	return rc;
+}
+
+int
+xtm_queue_notify_consumer(struct xtm_queue *queue)
+{
+	return notify_fd(queue->consumer_write_fd);
+}
+
+int
+xtm_queue_notify_producer(struct xtm_queue *queue)
+{
+	return notify_fd(queue->producer_write_fd);
+}
+
+int
+xtm_queue_probe(struct xtm_queue *queue)
+{
+	if (queue->queue.free_count() == 0) {
+		errno = ENOBUFS;
+		return -1;
+	}
+	return 0;
+}
+
+unsigned
+xtm_queue_count(struct xtm_queue *queue)
+{
+	return queue->queue.count();
+}
+
+int
+xtm_queue_push_fun(struct xtm_queue *queue, xtm_queue_fun_t fun,
+		   void *fun_arg)
+{
+	union xtm_msg xtm_msg;
+	xtm_msg.fun = fun;
+	xtm_msg.fun_arg = fun_arg;
+	if (queue->queue.put(&xtm_msg, 1) == 1)
+		return 0;
+	errno = ENOBUFS;
+	if (queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) {
```

Comment:
You might want to store the flags as booleans in the queue struct. It is not a public API and you don't need to calculate `&` every time.

Reply:
I thought about it, but in case we have more flags in future, it will be ugly to have 32 booleans.

Reply:
We won't have 32 flags here ever. But even if so, they could be bit fields as `bool flag : 1;` if the problem is in having 32 bytes for them, still better than using `& VERY_LONG_NAME_IN_CAPS` everytime. But whatever, fine to me.